contracts/account/paymaster/PaymasterERC20.sol

```diff
@@ -0,0 +1,101 @@
+    uint256 private constant POST_OP_COST = 30_000;
+    uint256 private constant POST_OP_COST_WITH_GUARANTOR = 50_000;
+
+    /// @inheritdoc PaymasterCore
+    function _validatePaymasterUserOp(
+        PackedUserOperation calldata userOp,
+        bytes32 userOpHash,
+        uint256 maxCost
+    ) internal virtual override returns (bytes memory context, uint256 validationData) {
+        (IERC20 token, uint48 validAfter, uint48 validUntil, uint256 tokenPrice, address guarantor) = _fetchDetails(
+            userOp.paymasterData()
+        );
+
+        uint256 prefundAmount = (maxCost +
+            (guarantor == address(0)).ternary(POST_OP_COST, POST_OP_COST_WITH_GUARANTOR) *
+            userOp.maxFeePerGas()).mulDiv(tokenPrice, _tokenPriceDenominator());
+
+        return (
+            abi.encodePacked(userOpHash, token, prefundAmount, tokenPrice, userOp.sender, guarantor),
+            token
+                .trySafeTransferFrom(guarantor == address(0) ? userOp.sender : guarantor, address(this), prefundAmount)
+                .packValidationData(validAfter, validUntil)
+        );
+    }
+
+    /// @inheritdoc PaymasterCore
+    function _postOp(
+        PostOpMode /* mode */,
+        bytes calldata context,
+        uint256 actualGasCost,
+        uint256 actualUserOpFeePerGas
+    ) internal virtual override {
+        bytes32 userOpHash = bytes32(context[0x00:0x20]);
+        IERC20 token = IERC20(address(bytes20(context[0x20:0x34])));
+        uint256 prefundAmount = uint256(bytes32(context[0x34:0x54]));
+        uint256 tokenPrice = uint256(bytes32(context[0x54:0x74]));
+        address user = address(bytes20(context[0x74:0x88]));
+        address guarantor = address(bytes20(context[0x88:0x9C]));
+
+        uint256 actualAmount = (actualGasCost +
+            (guarantor == address(0)).ternary(POST_OP_COST, POST_OP_COST_WITH_GUARANTOR) *
+            actualUserOpFeePerGas).mulDiv(tokenPrice, _tokenPriceDenominator());
+
+        if (guarantor == address(0)) {
+            token.safeTransfer(user, prefundAmount - actualAmount);
+            emit UserOperationSponsored(userOpHash, user, address(0), actualAmount, tokenPrice, false);
+        } else if (token.trySafeTransferFrom(user, address(this), actualAmount)) {
+            token.safeTransfer(guarantor, prefundAmount);
+            emit UserOperationSponsored(userOpHash, user, guarantor, actualAmount, tokenPrice, false);
+        } else {
+            token.safeTransfer(guarantor, prefundAmount - actualAmount);
+            emit UserOperationSponsored(userOpHash, user, guarantor, actualAmount, tokenPrice, true);
+        }
+    }
+
+    function _fetchDetails(
+        bytes calldata paymasterData
+    )
+        internal
+        view
+        virtual
+        returns (IERC20 token, uint48 validAfter, uint48 validUntil, uint256 tokenPrice, address guarantor);
+
+    function _tokenPriceDenominator() internal view virtual returns (uint256) {
+        return 1e6;
+    }
```

Comment:
Worth documenting this?

Reply:
Yes, this is missing some documentation!

One thing we could consider is to 
- remove that function
- put the denominator in the fetch details 
- pass it in the context.

Reply:
Could we not just require the necessary scaling to be done in the `_fetchDetails` function? Meaning, just always have a denominator of 1e18.